opal/corelib/kernel.rb

```diff
@@ -1006,21 +1006,38 @@ def warn(*strs)
 
   def raise(exception = undefined, string = undefined)
     %x{
-      if (exception == null && #$!) {
+      if (exception == null && #{$! != nil}) {
         throw #$!;
       }
-
       if (exception == null) {
         exception = #{RuntimeError.new};
       }
       else if (exception.$$is_string) {
         exception = #{RuntimeError.new exception};
       }
-      else if (exception.$$is_class) {
+      // using respond_to? and not an undefined check to avoid method_missing matching as true
+      else if (exception.$$is_class && #{exception.respond_to?(:exception)}) {
+        exception = #{exception.exception string};
+      }
+      else if (#{exception.kind_of?(Exception)}) {
+        // exception is fine
+      }
+      else if (exception.$$is_class && (exception === #{Exception} || #{exception < Exception})) {
```

Comment:
I think `exception < Exception` comparison [is not](http://opalrb.org/try/?code:class%20ApiException%20%3C%20Exception%0Aend%0A%0Aclass%20Dog%0Aend%0A%0Aputs%20ApiException%20%3C%20Exception%0Aputs%20Dog%20%3C%20Exception%0Aputs%20Exception%20%3C%20Exception%0A) necessary